src/hotspot/share/opto/addnode.cpp

```diff
@@ -769,6 +771,553 @@ uint AddPNode::match_edge(uint idx) const {
+  };
+
+private:
+  PhaseGVN* const      _phase;
+  Node*     const    _combine;
+  MemoryAdjacentStatus _order;
+  bool _require_reverse_bytes;    // Do we need add a ReverseBytes for merged load
+
+  NOT_PRODUCT( const CHeapBitMap &_trace_tags; )
+
+public:
+  MergePrimitiveLoads(PhaseGVN* phase, Node* combine) :
+    _phase(phase), _combine(combine), _order(Unknown), _require_reverse_bytes(false)
+    NOT_PRODUCT( COMMA _trace_tags(Compile::current()->directive()->trace_merge_loads_tags()) )
+  {}
+
+  // try merge adjacent loads, and return replace node for origin _combine
+  Node* run();
+
+private:
+  // Detect the embedding combine node is a candidate for merging loads
+  bool is_merged_load_candidate( ) const;
+  // Check other_load and load are compatible
+  bool is_compatible_load(const LoadNode* other_load, const LoadNode* load) const;
+  // From the seed load to collect load items for merging
+  void collect_merge_list(MergeLoadInfoList& merge_list, const LoadNode* load);
+  // Construct merge information item from input load
+  MergeLoadInfo* merge_load_info(LoadNode* load) const;
+  // Make the merged load and optional ReverseByte for replace
+  Node* make_merged_load(const MergeLoadInfoList& merge_list);
+
+  // Helper methods for merge loads optimization
+  static bool is_supported_load_opcode(int opcode);
+  static bool is_supported_combine_opcode(int opcode);
+  // Check the two combine operator Nodes can be reachable from one to another, and intermediate nodes has the same type
+  static bool is_reachable_combine_nodes(const Node* from, const Node* to);
+  MemoryAdjacentStatus get_adjacent_load_status(const LoadNode* first, const LoadNode* second) const;
+  // Go through ConvI2L which is unique output of input node
+  static const Node* by_pass_i2l(const Node* n);
+
+#ifndef PRODUCT
+  // Access to TraceMergeLoads tags
+  bool is_trace(TraceMergeLoads::Tag tag) const {
+    return _trace_tags.at(tag);
+  }
+
+  bool is_trace_basic() const {
+    return is_trace(TraceMergeLoads::Tag::BASIC);
+  }
+
+  bool is_trace_pointer_parsing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_PARSING);
+  }
+
+  bool is_trace_pointer_aliasing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ALIASING);
+  }
+
+  bool is_trace_pointer_adjacency() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ADJACENCY);
+  }
+
+  bool is_trace_success() const {
+    return is_trace(TraceMergeLoads::Tag::SUCCESS);
+  }
+#endif
+};
+
+bool MergePrimitiveLoads::is_supported_load_opcode(int opc) {
+  // Check for B/S/C/I
+  if (opc != Op_LoadB && opc != Op_LoadUB &&
+      opc != Op_LoadS && opc != Op_LoadUS &&
+      opc != Op_LoadI) {
+    return false;
+  }
+  return true;
+}
+
+bool MergePrimitiveLoads::is_supported_combine_opcode(int opc) {
+  return opc == Op_OrI || opc == Op_OrL;
+}
+
+// Go through ConvI2L which is unique output of input node
+const Node* MergePrimitiveLoads::by_pass_i2l(const Node* n) {
```

Comment:
```suggestion
const Node* MergePrimitiveLoads::bypass_ConvI2L(const Node* n) {
```
I think `bypass` is a single verb, not two wordy ;)

Reply:
Fixed